File: day1/main.py
left_list: list[int] = []
right_list: list[int] = []

with open("./puzzle_input.txt", "r") as file_input:
    line = file_input.readline().strip()
    while line is None or line != "":
        num_input = line.split(" ")
        left_list.append(int(num_input[0]))
        right_list.append(int(num_input[-1]))
        line = file_input.readline().strip()


def list_is_empty(li: list):
    return len(li) <= 0


def iter_total_distance(l_list: list[int], r_list: list[int]):
    total_distance = 0

    if len(l_list) != len(r_list):
        raise ValueError("List lengths do not match")

    l_list = sorted(l_list)
    r_list = sorted(r_list)

    for _ in range(len(l_list)):
        left, right = (l_list.pop(0), r_list.pop(0))
        total_distance += abs(left - right)

    return total_distance


iter_ans = iter_total_distance(left_list.copy(), right_list.copy())
print(iter_ans)


def similarity_score(l_list: list[int], r_list: list[int]):
    scores: dict[int, int] = {}

    def score(n: int):
        if n in scores.keys():
            return scores[n]

        scores[n] = n * r_list.count(n)
        return scores[n]

    similarity_score = 0
    for num in l_list:
        similarity_score += score(num)

    return similarity_score


s_score = similarity_score(left_list.copy(), right_list.copy())
print(s_score)

File: day1/Main.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Main {
	public static void main(String[] args) {
		File input = new File("puzzle_input.txt");
		List<Integer> leftList = new ArrayList<>();
		List<Integer> rightList = new ArrayList<>();
		getInput(input, leftList, rightList);

		int distance = calcTotalDistance(leftList, rightList);
		System.out.println(distance);

		int similarityScore = calcSimilarityScore(leftList, rightList);
		System.out.println(similarityScore);
	}

	private static final Map<Integer, Integer> SCORES = new HashMap<>();

	private static int askScores(int n, List<Integer> rList) {
		if (SCORES.containsKey(n))
			return SCORES.get(n);

		int score = (int) (n * rList.stream().filter(a -> a == n).count());
		SCORES.put(n, score);
		return score;
	}

	private static int calcSimilarityScore(List<Integer> leftList, List<Integer> rightList) {
		int similarityScore = 0;
		for (int value : leftList)
			similarityScore += askScores(value, rightList);

		return similarityScore;
	}

	public static void getInput(File input, List<Integer> l_list, List<Integer> r_list) {
		try (BufferedReader reader = new BufferedReader(new FileReader(input))) {
			while (true) {
				String line = reader.readLine();
				if (line == null)
					break;
				String[] split = line.trim().split(" ");
				l_list.add(Integer.parseInt(split[0]));
				r_list.add(Integer.parseInt(split[split.length - 1]));
			}

		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	public static int calcTotalDistance(List<Integer> l_list, List<Integer> r_list) {
		int totalDistance = 0;

		if (l_list.size() != r_list.size()) {
			throw new RuntimeException("List are different size");
		}

		List<Integer> lSorted = l_list.stream().sorted().collect(Collectors.toList());
		List<Integer> rSorted = r_list.stream().sorted().collect(Collectors.toList());

		for (int i = 0; i < lSorted.size(); i++) {
			int left = lSorted.get(i);
			int right = rSorted.get(i);

			totalDistance += Math.abs(left - right);
		}

		return totalDistance;
	}
}